prusti-interface/src/environment/tymap.rs

```diff
@@ -0,0 +1,130 @@
+use super::{traits, Environment};
+use crate::data::ProcedureDefId;
+use log::{trace, warn};
+use rustc_hash::{FxHashMap, FxHashSet};
+use rustc_middle::{ty, ty::subst::SubstsRef};
+
+type Key<'tcx> = ty::Ty<'tcx>;
+type Value<'tcx> = ty::Ty<'tcx>;
+
+#[derive(Default, Clone, Debug)]
+pub struct SubstMap<'tcx> {
+    // TODO: Make this non-public
+    pub map: FxHashMap<Key<'tcx>, Value<'tcx>>,
+}
+
+impl<'tcx> SubstMap<'tcx> {
+    // TODO hansenj: Docs
+    // TODO hansenj: Restructure, avoid dup code
+    // TODO hansenj: Debugging
+    pub fn build<'a>(
+        env: &'a Environment<'tcx>,
+        def_id: ProcedureDefId,
+        call_substs: SubstsRef<'tcx>,
+    ) -> Self {
+        trace!(
+            "Building tymap for {:?} with substs {:?}",
+            def_id,
+            call_substs
+        );
+
+        let mut tymap = Self::default();
+
+        // Add generic information from the call to this method
+        let own_substs = ty::List::identity_for_item(env.tcx(), def_id);
+        own_substs
+            .iter()
+            .zip(call_substs)
+            .for_each(|(generic_arg, call_arg)| tymap.insert_arg(generic_arg, call_arg));
+
+        // In case this is a trait method implementation, we also extend
+        // tymap with information from the implementation itself
+        let maybe_impl_did = if call_substs.is_empty() {
+            None
+        } else {
+            env.find_impl_of_trait_method_call(def_id, call_substs)
+        };
+
+        if let Some(impl_did) = maybe_impl_did {
+            let impl_own_substs = ty::List::identity_for_item(env.tcx(), impl_did);
+            let trait_did = env.tcx().trait_of_item(def_id).unwrap();
+            let trait_ref = ty::TraitRef::from_method(env.tcx(), trait_did, call_substs);
+            let trait_binder = ty::Binder::bind_with_vars(trait_ref, ty::List::empty());
```

Comment:
Is this the same as `ty::Binder::dummy(trait_ref)`?

Reply:
This seems to be the proper choice when I read the docs :+1: